### ISSUE TITLE
feature request: install revolt without docker

### ISSUE THREAD

**Issue by sysadminpower2019** [revoltchat/self-hosted#91]
### What do you want to see?

Hello,

Would it be possible to install revoltchat without using docker? This would be important for smaller instances and machines with lower specs

**Comment by insertish** [comment]
Are you asking for a guide or if it's possible?

Build instructions are detailed on individual projects, e.g. https://github.com/revoltchat/backend#development-guide
(if you pin to a tag, the development guide is functionally the same steps you'd take to deploy production)

**Comment by sysadminpower2019** [comment]
yes the ideal would be a guide on how to run this docker freee. the link you provided still uses docker as part of the documentation, so I don't really understand how you can use this to run a dockerless instance

**Comment by insertish** [comment]
> yes the ideal would be a guide on how to run this docker freee. the link you provided still uses docker as part of the documentation, so I don't really understand how you can use this to run a dockerless instance

Apologies, didn't realise.
I'm not sure whether this is in scope for the project, especially since maintaining all of these services separately can become cumbersome very quickly. Is there any reason to not use a set of containers here?

**Comment by alexpyattaev** [comment]
Containers are fairly heavy on the CPU and especially on the disk usage. So on a small VPS this could be a serious problem.

**Comment by tuaris** [comment]
> Is there any reason to not use a set of containers here?

I would also like to add that this limits the platform choice to only operating systems that can run Docker.  Not everything Unix is Linux :)

**Comment by rrimc69** [comment]
yes please how do I install a Revolt server, without the docker requirement? without opening container warz wounds, docker is not the be all and end all to C group isolation. 

As Tuaris points out, not all *nix have or package docker, and many of us do not want, or need the overhead 'on-box' . . . I'm part way thru trying to un-dockerise the documentation, which is not so fun. It may be all you need is a 'non docker' howto for your existing code .. if that's at all possible. 

Hell might try drop it in a Solaris zone .. definitely no docker in there.  *BSD jails ? Lots of reasons one might not want docker.

**Comment by eekhof** [comment]
I think providing the information that @tuaris asked for in a concise manner here would already help immensely 👍

**Comment by ghost** [comment]
Despite their claim of being user focused, it seems end user friendliness for this project is very low, and it seems they are only committed to a re-branding that has dropped privacy as a headline feature, and accelerating overburdening their supposed server with pointless containerised middleman complexity garbage. 

I'm looking elsewhere for a private lightweight discord alternative, as apathy on this topic, along with  mandating containers and refusing to document alternatives installation options, demonstrates for me, this is not it. 

If you find this from the internet, think twice.

**Comment by sinus-x** [comment]
>  mandating containers and refusing to document alternatives

Maintenance of even just the documentation of non-container use-cases would be significant. This repository does accept pull requests, you can propose a community maintained version of the docs.

Containers add essentially zero CPU overhead, they are just a user-friendly wrapper around process namespaces and cgroups, provided by kernel, with additional networking and service management on top.

It is fair containers need additional disk space compared to OCI-less solution, but... all 1st party Stoat containers take up 1255 MB, which is less than the rest of supporting containers (1542 MB). And if you plan to allow attachments in your instance, they will outgrow the infrastructure stack size very fast.

If you are scared of 2.7 GB of OCI images, I really do think this isn't for you, and you'd be better off running Mattermost, XMPP or IRC.

**Comment by alexbobp** [comment]
I feel the need to respond to what I see are some troublesome arguments.  First of all, you're right that the CPU overhead of docker is trivial, and the memory/disk overhead is probably fine for anyone looking to do a nontrivial stoat deployment.  To me, that's not the issue.

The issue is that deployment instructions are what give sysadmins and developers an understanding of the tech stack and how to work with it.  This is important institutional knowledge that you are failing to transfer if all you provide is "use the docker image" which is relatively opaque.  To wit: you are inviting volunteers to write the deployment documentation, but how would the volunteers figure out how to deploy stoat in order to write said documentation?  It's undocumented!

I would analogize this to when you find a project that is "open source" but does not bother to document how to compile it, and conveniently tells you that you shouldn't mind, because their precompiled binaries work great.  Docker images are like the precompiled binaries of sysadmin work, and yes they are a handy deployment option, but if the project wants to gain contributors, or simply the *trust* of potential users who want to understand something before they dive in, then there is important institutional knowledge here that is not being transferred.

Edit: Do you have any documentation on how your docker images are built?  That would be a very good place to start.  If we know how the stoat docker images themselves are built, then that plus the docker compose file I think provides enough information that someone could reverse engineer deployment instructions.

**Comment by DeclanChidlow** [comment]
@alexbobp Most of what you're discussing is covered in the previously mentioned instructions for individual projects. For example, the back-end: https://github.com/stoatchat/stoatchat#development-guide

As mentioned, people are welcome to make user-friendly documentation for end-user deployments using the existing documentation for compiling and such elsewhere as a base, but we currently just don't have the bandwidth to maintain complex additional documentation.

**Comment by sinus-x** [comment]
Disclaimer: I am not a maintainer of Stoat in any capacity, I am a passer by just as you are.

> how would the volunteers figure out how to deploy stoat in order to write said documentation? It's undocumented!

This repository contains `compose.yml`, which is the literal recipe for machines to bring the services up. That is the documentation, together with the literal source code in the stoatchat/stoatchat repository.

> the project wants to gain contributors, or simply the trust of potential users who want to understand something before they dive in

If you aren't willing to make the first step and get your hands dirty, what kind of contributions do you expect to provide?

> Do you have any documentation on how your docker images are built?

To find https://github.com/stoatchat/stoatchat/blob/main/.github/workflows/docker.yaml took me literal one minute.

**Comment by alexbobp** [comment]
Maybe it's my lack of expertise in docker.  I was definitely searching for more than a minute.  Apologies if I missed the obvious.  Clear I'm not the one who will be qualified to reverse engineer this into documentation...

**Comment by jameshiller** [comment]
Docker trivializes deployment. It would be more complicated for everyone involved to spend time hand-holding alternative deployment options. If someone truly has a requirement to not use Docker and they've made an effective technical assessment that a nonstandard deployment is more appropriate, they're probably technical to figure out how to manually deploy it given all the pieces are in the repos. Anything less is just people unwilling to spend a little time understanding docker (and how great and easy it can be when used correctly).

**Comment by neuza-git** [comment]
> 
> To find https://github.com/stoatchat/stoatchat/blob/main/.github/workflows/docker.yaml took me literal one minute.

that's a yaml workflow, it doesn't contain any of the steps required to setup stoat, that could be one of the reason "it took you one minute" because you didn't even take the time to take a look at the link you sent

**Comment by sinus-x** [comment]
> you didn't even take the time to take a look at the link you sent

The question was "how your docker images are built", not "how the software stack should get deployed".

> it doesn't contain any of the steps required to setup stoat

To set up Stoat without containers, read through [compose.yml](https://github.com/stoatchat/self-hosted/blob/main/compose.yml) and install the dependencies manually: MongoDB, Redis/KeyDB/Valkey, RabbitMQ, MinIO/Garage/SeaweedFS, build the Rust binaries contained in stoatchat/stoatchat (`mise build`, but `cargo build` would work as well I believe), create `Revolt.toml` [in the current directory](https://github.com/stoatchat/stoatchat/blob/24d0d2b7266f6f8a692d0a52704acfecf517674c/crates/core/config/src/lib.rs#L56-L63), fill in the configuration (manually or by running [setup script](https://github.com/stoatchat/self-hosted/blob/main/generate_config.sh)) and run all the services (either as built binaries, or if you changed the code, you can even do `bash .mise/tasks/service/<tab> &`).

The last step is to set up Caddy or other reverse proxy to properly route the traffic to the appropriate backend services, example Caddyfile is [available](https://github.com/stoatchat/self-hosted/blob/main/Caddyfile).

...

Hey, I don't want to fight you. All I am trying to say is the source code for everything is available for you to read. Clone the repository and ask Cursor or Claude if you feel you can't read YAML or Rust.

Kicking around and complaining that not providing deployment documentation in a technology of _your choice_ is "apathy" and "low end user friendliness" is not a way to go.

A side note: you are [legally not allowed](https://github.com/stoatchat/stoatchat/blob/24d0d2b7266f6f8a692d0a52704acfecf517674c/LICENSE#L590-L599) to demand anything of this project.
